hal/src/nRF52840/exflash_hal.c

```diff
@@ -585,27 +554,33 @@ int hal_exflash_special_command(hal_exflash_special_sector_t sp, hal_exflash_com
 int hal_exflash_sleep(bool sleep, void* reserved) {
     hal_exflash_lock();
     if (sleep) {
-        // Suspend I2C
+        // Suspend external flash
         if (qspi_state != HAL_EXFLASH_STATE_ENABLED) {
             hal_exflash_unlock();
             return SYSTEM_ERROR_INVALID_STATE;
         }
         // Put external flash into sleep and disable QSPI peripheral
-        if (hal_exflash_special_command(HAL_EXFLASH_SPECIAL_SECTOR_NONE, HAL_EXFLASH_COMMAND_SLEEP, NULL, NULL, 0) != SYSTEM_ERROR_NONE ||
-            hal_exflash_uninit() != SYSTEM_ERROR_NONE) {
-            hal_exflash_unlock();
-            return SYSTEM_ERROR_INTERNAL;
+        if (hal_exflash_special_command(HAL_EXFLASH_SPECIAL_SECTOR_NONE, HAL_EXFLASH_COMMAND_SLEEP, NULL, NULL, 0)) {
+            // It may fail as the flash chip might be in an ongoing program/erase operation.
+            // Suspend it.
+            int ret = hal_exflash_special_command(HAL_EXFLASH_COMMAND_NONE, HAL_EXFLASH_COMMAND_SUSPEND_PGMERS, NULL, NULL, 0);
```

Comment:
If we get here we are not putting the QSPI flash back into sleep state though.

Reply:
Just looked through the datasheet and found that the sleep command is not acceptable when program/erase is suspended. So if the first attempt executing the sleep command failed, we should return, leaving the external flash running.

Reply:
We can always abort the pending write/erase as well. It shouldn't cause any problems.

Reply:
So simply re-initialize the external flash if executing the sleep command failed and then re-execute the sleep command to put it into sleep state?

Reply:
I'd probably suggest:

1. Sleep command
2. If fails suspend write/erase
3. QSPI_STD_CMD_RSTEN (which will abort suspended operation)
4. Sleep
5. If fails, give up probably

Reply:
Good suggestion! Will do.